### ISSUE TITLE
Storage CI/CD tests failing

### ISSUE THREAD

**Issue by antkmsft** [Azure/azure-sdk-for-cpp#224]
CI/CD is blocked because storage tests are failing:
```
Test command: /home/vsts/work/1/s/build/sdk/storage/test/azure-storage-test
2: Test timeout computed to be: 10000000
2: [==========] Running 17 tests from 5 test suites.
2: [----------] Global test environment set-up.
2: [----------] 1 test from BlobServiceClientTest
2: [ RUN      ] BlobServiceClientTest.ListContainers
2: unknown file: Failure
2: C++ exception with description "basic_string::_M_construct null not valid" thrown in the test fixture's constructor.
2: [  FAILED  ] BlobServiceClientTest.ListContainers (0 ms)
2: [----------] 1 test from BlobServiceClientTest (0 ms total)
2: 
2: [----------] 4 tests from BlobContainerClientTest
2: unknown file: Failure
2: C++ exception with description "basic_string::_M_construct null not valid" thrown in SetUpTestSuite().
2: [ RUN      ] BlobContainerClientTest.CreateDelete
2: unknown file: Failure
2: C++ exception with description "basic_string::_M_construct null not valid" thrown in the test body.
2: [  FAILED  ] BlobContainerClientTest.CreateDelete (0 ms)
2: [ RUN      ] BlobContainerClientTest.Metadata
1/1 Test #2: azure-storage-test ...............***Exception: SegFault  0.15 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.22 sec

The following tests FAILED:
	  2 - azure-storage-test (SEGFAULT)
Errors while running CTest
```

I have to comment out the test to unblock my and @vhvb1989's PRs.

I think it is in BlobServiceClientTest's constructor. Null connection string gets returned, and it fails. I could've safeguard the code from null cstr myself, but I think it won't go much further. It looks like it expects connection string being set as an environment variable, and when it is not set on the CI/CD machine, it won't be able to go much further, i.e. the test would still fail, it looks like.

**Comment by JinmingHu-MSFT** [comment]
Well, actually I don't know that I should do regarding this issue.

The problem is, Azure Core team made some breaking changes, which made Storage unit tests not work anymore. I don't mean Core team cannot make any breaking changes. But in turn, you cannot expect Storage code to work all the time no matter how Azure Core API changes. Because Storage team can never anticipate what kind of breaking changes you're going to make in Azure Core SDK.

**Comment by antkmsft** [comment]
+ @vhvb1989

**Comment by JinmingHu-MSFT** [comment]
And now, there are more than one PR https://github.com/Azure/azure-sdk-for-cpp/pull/214 https://github.com/Azure/azure-sdk-for-cpp/pull/212 that fail because of  this, and Core team is making all kinds of changes to storage code just to make CI pass. Those changes even have conflicts. I don't know what I'm supposed to do here.

Suppose in the future, there's going to be dozens of other services besides storage service. Are you going to make changes for ALL those services every time you introduce a breaking change in Azure Core SDK? I don't think that's practicable. I think we should figure out a way to deal with this situation.

**Comment by antkmsft** [comment]
I think we should work together to improve this. On your part, your tests were not running as a part of CI/CD, correct? So, it probably was hard to know when we would break you. Quite honestly, storage is a part of the project now, and I personally have problem building the full SDK project now, on my local machine (libxml2 not found) - and there are no instructions and readmes on how to make it pass. So, my PR did not affect Storage at all, so when I was building locally, I just disabled Storage from building (but CI/CD did built it for me).
Right now, I have to disable these tests, because master is broken, and that should never happen. The changes I am checking in are unrelated to both Core or Storage.

**Comment by JinmingHu-MSFT** [comment]
@antkmsft Yes, I totally agree that we should work together to improve this.

I want to clarify that Storage test cases were able to build and run in the past, but then Core SDK introduced some breaking changes and Storage test cases didn't work anymore. Even before storage team was able to fix the build error, those breaking changes was merged into master, with some bugs in it. Then up until now, storage team are never able to fix the CI because those bugs are not in storage code, they're in Core code.

Then about the dependency issue, I think it would be best for no matter ourselves as developers or our customers to install dependencies in the same way. I didn't see any essential difference between libxml2 and libcurl as external dependencies. So if you install libcurl via vcpkg, it would also work to install libxml2 via vcpkg. If you install libcurl from source, it also works to install libxml2 from source. I agree with you that we should add some instructions about how to install dependencies. Is there any instructions or readme about how to install libcurl that we can refer to when writing instructions of how to install libxml2?

Right now, i'm also okay if you want to disable all storage test cases to get CI pass. We're going to fix them all and add them back after all bugs in Core SDK are fixed. But I still think it's necessary for us to think about how to handle situations like this in the future. I mean sooner or later, there's going to be another breaking change in Core SDK, and when we have many many services, most of which (if not all) depend on Core SDK, all those services will be broken.

**Comment by antkmsft** [comment]
I see absolutely nothing to disagree on at the very least least in essence, and I think all of us want to see it fixed, integration strengthened, and the situation never happening again. If there's anything I can help with, please let me know.
You are right about the libxml2, I haven't thought it could (and should) be installed the same way. My slight personal problem is that I installed libcurl a long ago and I don't remember how to do it, but I will figure out :) C repo should have docs for that.

**Comment by vhvb1989** [comment]
>I want to clarify that Storage test cases were able to build and run in the past, but then Core SDK introduced some breaking changes and Storage test cases didn't work anymore. 

We just realized last Friday that our CI was not compiling/running any test at all (compilation flag -DBUILD_TESTING=ON was missing). Neither Core or Storage.  We had another breaking change with the Nullable feature and then we found out about some time ago there was an accidentally typo in CI scripts that disabled tests. This was the root cause of that break storage team had (and we were wondering about it because CI was passing).

So, After fixing CI, the real status of Master came up!. Error while compiling storage CI tests and also trying to run.

I worked las week to make storage Unit tests work in my local env.  Then I created my PR for unique_ptr<StreamBody> and I got storage team's help to make that PR ready to take our master to stable.

I don't think we should disable Unit tests from storage.  We need to always compile them and if we don't want to run them, we can use `ctest -R core -V` to run only Core unit test. @antkmsft I can help with dependencies installation and also on how to run Storage unite tests locally (all you need is a connection string to an storage account).

So, I propose to marge my PR and then I can help to rebase any other branch that needs to be merged as well

**Comment by antkmsft** [comment]
https://github.com/Azure/azure-sdk-for-cpp/pull/227

**Comment by RickWinter** [comment]
@JinmingHu-MSFT It was not that core made changes that broke storage tests.  The Ops team discovered that Storage tests were not being built/run on all platforms.  We enabled the pipelines to run as they should have been all along.  This resulted in the pipelines flagging that the storage tests had errors.  It was the right thing to do to enable the pipelines to prevent further breaks from occurring.

**Comment by antkmsft** [comment]
They also compile with warnings treated as errors. And two - they seem to be live tests, not the isolated unit tests? I.e. they require storage connection, so how the Azure resources needed to run the test should look like? Would just any storage connection work?

**Comment by JinmingHu-MSFT** [comment]
@RickWinter @antkmsft @vhvb1989 Things have got kind of complex here. Let's not discuss what happened in the past anymore.

Now I've fixed everything in storage code. That means no build errors, no compiler warnings that will be treated as errors. Here's the Azure Pipelines result https://github.com/Azure/azure-sdk-for-cpp/runs/821394831.

Storage tests can run in my local environment. There are two obstacles to making unit tests work on Azure Pipelines.
1. storage account credential.
2. The remaining two bugs in HTTP client. https://github.com/Azure/azure-sdk-for-cpp/issues/198

**Comment by JinmingHu-MSFT** [comment]
@antkmsft Here's how to create a storage account for tests.

1. First create a General Purpose V2 storage account.
![create](https://user-images.githubusercontent.com/52266153/86134486-4471ab80-bb1c-11ea-92f6-5073da6e157b.png)

2. If it's for ADLS Gen2 test, enable HNS. Since ADLS Gen2 unit tests are not checked in yet, so you can skip this.
![adls](https://user-images.githubusercontent.com/52266153/86134625-73881d00-bb1c-11ea-9c2f-659d21a2c690.png)

3. Enable Soft Delete Blob
![softdelete](https://user-images.githubusercontent.com/52266153/86134684-83076600-bb1c-11ea-95c4-cfd6efac5bf7.png)

4. Copy the connection string
![key](https://user-images.githubusercontent.com/52266153/86135304-32443d00-bb1d-11ea-9e90-05ca667962b0.png)

5. Put the connection string [here](https://github.com/Azure/azure-sdk-for-cpp/blob/74ce80659be3dd5e70539c6711f3f901f38221ce/sdk/storage/test/test_base.cpp#L19) or set environment variable `STANDARD_STORAGE_CONNECTION_STRING`.

**Comment by danieljurek** [comment]
When it comes time to implement live testing see here for guidance on how to automate creation/configuration of storage accounts using ARM templates: https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/48/Create-a-new-Live-Test-pipeline?anchor=test-resources.json

EngSys will provide the harness that deploys and cleans up the resources provided in the ARM templates tracked in https://github.com/Azure/azure-sdk-for-cpp/issues/166

**Comment by danieljurek** [comment]
Estimating 3 complexity for buffer space to iterate on any late breaking errors. Adding in the provision and deprovision steps is trivial but errors sometimes crop up in this process.

**Comment by Jinming-Hu** [comment]
CI/CD has been working for some time. Close this issue.
